core/src/main/scala/kafka/log/Log.scala

```diff
@@ -582,59 +583,42 @@ class Log(val dir: File,
    * `NOTE:` OffsetRequest V0 does not use this method, the behavior of OffsetRequest V0 remains the same as before
    * , i.e. it only gives back the timestamp based on the last modification time of the log segments.
    *
-   * @param timestamp The given timestamp for offset fetching.
+   * @param targetTimestamp The given timestamp for offset fetching.
    * @return The offset of the first message whose timestamp is greater than or equals to the given timestamp.
+   *         None if no such message is found.
    */
-  def fetchOffsetsByTimestamp(timestamp: Long): Long = {
-    debug(s"Searching offset for timestamp $timestamp")
-    val segsArray = logSegments.toArray
-    if (timestamp == OffsetRequest.EarliestTime)
-      return segsArray(0).baseOffset
+  def fetchOffsetsByTimestamp(targetTimestamp: Long): Option[TimestampOffset] = {
+    debug(s"Searching offset for timestamp $targetTimestamp")
 
-    // set the target timestamp to be Long.MaxValue if we need to find from the latest.
-    val targetTimestamp = timestamp match {
-      case OffsetRequest.LatestTime => Long.MaxValue
-      case _ => timestamp
-    }
+    if (config.messageFormatVersion.compare(KAFKA_0_10_0_IV0) < 0 &&
+        targetTimestamp != ListOffsetRequest.EARLIEST_TIMESTAMP &&
+        targetTimestamp != ListOffsetRequest.LATEST_TIMESTAMP)
+      throw new InvalidRequestException(s"Cannot search offsets based on timestamp because message format version " +
+          s"for partition $topicAndPartition is ${config.messageFormatVersion} which is earlier than the minimum " +
+          s"required version $KAFKA_0_10_0_IV0")
 
-    var foundOffset: Long = -1L
-    // We have this while loop here to make sure we are returning the valid offsets to our best knowledge.
-    // This while loop is to handle the case where the log is truncated during the timestamp search and we did not
-    // find any message. In this case, we need to retry the search.
-    do {
-      val targetSeg = {
-        // Get all the segments whose largest timestamp is smaller than target timestamp
-        val earlierSegs = segsArray.takeWhile(_.largestTimestamp < targetTimestamp)
-        // We need to search the first segment whose largest timestamp is greater than the target timestamp if there is one.
-        if (earlierSegs.length < segsArray.length)
-          segsArray(earlierSegs.length)
-        else
-          earlierSegs.last
-      }
+    // For the earliest and latest, we do not need to return the timestamp.
```

Comment:
Thinking about this a little more... It seems like the inability to fetch the earliest or latest timestamp is a somewhat big drawback of these APIs. Here's a couple questions:

1. The first question is the signature:
```java
Map<TopicPartition, Long> earliestOffsets(Set<TopicPartition> partitions);
```
You're saying we can't return `OffsetAndTimestamp` because the timestamp is not supported for old versions. But we could always use `NO_TIMESTAMP` in that case like we do for the timestamp field in `ConsumerRecord`, right?

2. In the case of `latestOffset()`, we actually return the high watermark, and it makes sense that there's no timestamp associated with that. First, is the name misleading? Should it be called `highWatermark()` instead if that's actually what it returns? Alternatively, I'm wondering if it would be more useful in practice to have this method return the offset and timestamp of the last produced message.